README
NAME
    Template::Tiny - Template Toolkit reimplemented in as little code as
    possible

SYNOPSIS
      my $template = Template::Tiny->new(
          TRIM => 1,
      );
      
  # Print the template results to STDOUT
      $template->process( <<'END_TEMPLATE', { foo => 'World' } );
      Hello [% foo %]!
      END_TEMPLATE

DESCRIPTION
    WARNING: THIS MODULE IS EXPERIMENTAL AND SUBJECT TO CHANGE WITHOUT
    NOTICE

    YOU HAVE BEEN WARNED!

    Template::Tiny is a reimplementation of a partial subset of the Template
    Toolkit in as few lines of code as possible.

    It is intended for use in light-usage, low-memory, or low-cpu templating
    situations, where you may need to upgrade to the full feature set in the
    future, or if you want the familiarity of TT-style templates.

    It is intended to have fully-compatible template and stash usage, with a
    limited by similar Perl API.

    Unlike Template Toolkit, Template::Tiny will process templates without a
    compile phase (but despite this is still quicker, owing to heavy use of
    the Perl regular expression engine.

  SUPPORTED USAGE
    Only the default "[% %]" tag style is supported.

    Both the [%+ +%] style explicit whitespace and the [%- -%] style
    explicit chomp are support, although the [%+ +%] version is unneeded as
    Template::Tiny does not support default-enabled PRE_CHOMP or POST_CHOMP.

    Variable expressions in the form foo.bar.baz are supported.

    Appropriate simple behaviours for ARRAY reference, HASH reference and
    objects are supported, but not "VMethods" such as array lengths.

    IF, ELSE and UNLESS conditions are supported, but only with simple
    foo.bar.baz conditions.

    Support for looping is available, in the most simple [% FOREACH item IN
    list %] form.

    All four IF/ELSE/UNLESS/FOREACH control structures are able to be nested
    to arbitrary depth.

    The treatment of "_private" hash and method keys is compatible with
    Template Toolkit, returning null or false rather than the actual content
    of the hash key or method.

    Anything beyond this is currently out of scope

METHODS
  new
      my $template = Template::Tiny->new(
          TRIM => 1,
      );

    The "new" constructor is provided for compatibility with Template
    Toolkit.

    The only parameter it currently supports is "TRIM" (which removes
    leading and trailing whitespace from processed templates).

    Additional parameters can be provided without error, but will be
    ignored.

  process
      # DEPRECATED: Return template results
      my $text = $template->process( \$input, $vars );
      
  # Print template results to STDOUT
      $template->process( \$input, $vars );
      
  # Generate template results into a variable
      my $output = '';
      $template->process( \$input, $vars, \$output );

    The "process" method is called to process a template.

    The first parameter is a reference to a text string containing the
    template text. A reference to a hash may be passed as the second
    parameter containing definitions of template variables.

    If a third parameter is provided, it must be a scalar reference to be
    populated with the output of the template.

    For a limited amount of time, the old deprecated interface will continue
    to be supported. If "process" is called without a third parameter, and
    in scalar or list contest, the template results will be returned to the
    caller.

    If "process" is called without a third parameter, and in void context,
    the template results will be print()ed to the currently selected file
    handle (probably "STDOUT") for compatibility with Template.

SUPPORT
    Bugs should be reported via the CPAN bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Template-Tiny>

    For other issues, or commercial enhancement or support, contact the
    author.

AUTHOR
    Adam Kennedy <adamk@cpan.org>

SEE ALSO
    Config::Simple

COPYRIGHT
    Copyright 2009 - 2010 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Changes for Perl extension Template-Tiny

0.11 Mon 22 Feb 2010
	- Adding preparse-specific test script
	- Add compatibility with the horrible TT process()  signature (MSTROUT)

0.10 Sun 24 Jan 2010
	- The use of the constructor is now compulsory, as per Template.pm
	- Adding support for constructor params
	- Adding support for constructor TRIM param
	- The process method now works in both static and instance modes

0.09 Wed 16 Dec 2009
	- Adding support for "private" _foo keys

0.08 Sun 13 Dec 2009
	- Removing some minor code no longer needed.
	- Updating the POD documentation

0.07 Sun 13 Dec 2009
	- Recursion is now processed top to bottom
	- Added support for FOREACH

0.06 Sun 13 Dec 2009
	- Added support for nested [% IF %] and [% UNLESS %] conditions

0.05 Thu 10 Dec 2009
	- Expression chains in a condition that terminate in an object
	  will boolify the object instead of stringifying it + tests for it

0.04 Thu 10 Dec 2009
	- Reduced implementation by 8kb
	- As a side effect of this, added support for UNLESS/ELSE
	- Matching behaviour of TT to NOT chomp on -%](space)(EOF)
	- Switch to Tiny-style Makefile.PL

0.03 Thu 10 Dec 2009
	- Bug fix for methods in expressions
	- Added test case for method calls

0.02 Tue  8 Dec 2009
	- Added support for [%- -%]
	- Added support for recursive simple expressions
	- Added support for [% IF expr %]
	- Added support for [% UNLESS expr %]
	- Added support for [% ELSE %]

0.01 Sun  6 Dec 2009
	- Created initial release


Makefile.PL
use strict;
use vars qw{$VERSION};
BEGIN {
	require 5.003_96;
	$VERSION = '0.11';
}
use ExtUtils::MakeMaker;

WriteMakefile1(
	MIN_PERL_VERSION => '5.00503', # from Tiny.pm
	META_MERGE => {
		resources => {
			repository => 'http://svn.ali.as/cpan/trunk/Template-Tiny',
		},
	},
	BUILD_REQUIRES => {
		# Skip on Windows to avoid breaking ActivePerl PPMs
		# 0.47 means 5.6.2 or newer, which everyone on Win32 has.
		($^O eq 'MSWin32' ? () : ('Test::More' => '0.47')),
		'Capture::Tiny' => '0.07',
	},
	NAME      => 'Template::Tiny',
	ABSTRACT  => 'Template Toolkit reimplemented in as little code as possible',
	VERSION   => $VERSION,
	PREREQ_PM => { },
	AUTHOR    => 'Adam Kennedy <adamk@cpan.org>',
	LICENSE   => 'perl',
);

# Written by Alexandr Ciornii, version 0.20. Added by eumm-upgrade.
sub WriteMakefile1 {
	my %params = @_;
	my $eumm_version = $ExtUtils::MakeMaker::VERSION;
	$eumm_version = eval $eumm_version;
	die "License not specified" if not exists $params{LICENSE};
	if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
		#EUMM 6.5502 has problems with BUILD_REQUIRES
		$params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
		delete $params{BUILD_REQUIRES};
	}
	delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
	delete $params{META_MERGE} if $eumm_version < 6.46;
	delete $params{LICENSE} if $eumm_version < 6.31;
	delete $params{AUTHOR} if $] < 5.005;

	WriteMakefile(%params);
}


lib/Template/Tiny.pm
package Template::Tiny;

# Load overhead: 40k

use 5.00503;
use strict;

$Template::Tiny::VERSION = '0.10';

# Evaluatable expression
my $EXPR = qr/ [a-z_][\w.]* /xs;

# Opening [% tag including whitespace chomping rules
my $LEFT = qr/
	(?:
		(?: (?:^|\n) [ \t]* )? \[\%\-
		|
		\[\% \+?
	) \s*
/xs;

# Closing %] tag including whitespace chomping rules
my $RIGHT  = qr/
	\s* (?:
		\+? \%\]
		|
		\-\%\] (?: [ \t]* \n )?
	)
/xs;

# Preparsing run for nesting tags
my $PREPARSE = qr/
	$LEFT ( IF | UNLESS | FOREACH ) \s+
		(
			(?: \S+ \s+ IN \s+ )?
		\S+ )
	$RIGHT
	(?!
		.*?
		$LEFT (?: IF | UNLESS | FOREACH ) \b
	)
	( .*? )
	(?:
		$LEFT ELSE $RIGHT
		(?!
			.*?
			$LEFT (?: IF | UNLESS | FOREACH ) \b
		)
		( .+? )
	)?
	$LEFT END $RIGHT
/xs;

# Condition set
my $CONDITION = qr/
	\[\%\s
		( ([IUF])\d+ ) \s+
		(?:
			([a-z]\w*) \s+ IN \s+
		)?
		( $EXPR )
	\s\%\]
	( .*? )
	(?:
		\[\%\s \1 \s\%\]
		( .+? )
	)?
	\[\%\s \1 \s\%\]
/xs;

sub new {
	bless { @_[1..$#_] }, $_[0];
}

sub process {
	my $self  = shift;
	my $copy  = ${shift()};
	my $stash = shift || {};

	local $@  = '';
	local $^W = 0;

	# Preprocess to establish unique matching tag sets
	my $id = 0;
	1 while $copy =~ s/
		$PREPARSE
	/
		my $tag = substr($1, 0, 1) . ++$id;
		"\[\% $tag $2 \%\]$3\[\% $tag \%\]"
		. (defined($4) ? "$4\[\% $tag \%\]" : '');
	/sex;

	# Process down the nested tree of conditions
	$self->_process( $stash, $copy );
}

sub _process {
	my ($self, $stash, $text) = @_;

	$text =~ s/
		$CONDITION
	/
		($2 eq 'F')
			? $self->_foreach($stash, $3, $4, $5)
			: eval {
				$2 eq 'U'
				xor
				!! # Force boolification
				$self->_expression($stash, $4)
			}
				? $self->_process($stash, $5)
				: $self->_process($stash, $6)
	/gsex;

	# Resolve expressions
	$text =~ s/
		$LEFT ( $EXPR ) $RIGHT
	/
		eval {
			$self->_expression($stash, $1)
			. '' # Force stringification
		}
	/gsex;

	# Trim the document
	$text =~ s/^\s*(.+?)\s*\z/$1/s if $self->{TRIM};

	return $text;
}

# Special handling for foreach
sub _foreach {
	my ($self, $stash, $term, $expr, $text) = @_;

	# Resolve the expression
	my $list = $self->_expression($stash, $expr);
	unless ( ref $list eq 'ARRAY' ) {
		return '';
	}

	# Iterate
	return join '', map {
		$self->_process( { %$stash, $term => $_ }, $text )
	} @$list;
}

# Evaluates a stash expression
sub _expression {
	my $cursor = $_[1];
	my @path   = split /\./, $_[2];
	foreach ( @path ) {
		# Support for private keys
		return undef if substr($_, 0, 1) eq '_';

		# Split by data type
		my $type = ref $cursor;
		if ( $type eq 'ARRAY' ) {
			return '' unless /^(?:0|[0-9]\d*)\z/;
			$cursor = $cursor->[$_];
		} elsif ( $type eq 'HASH' ) {
			$cursor = $cursor->{$_};
		} elsif ( $type ) {
			$cursor = $cursor->$_();
		} else {
			return '';
		}
	}
	return $cursor;
}

1;

__END__

=pod

=head1 NAME

Template::Tiny - Template Toolkit reimplemented in as little code as possible

=head1 SYNOPSIS

  my $template = Template::Tiny->new(
      TRIM => 1,
  );
  
  $template->process( <<'END_TEMPLATE', { foo => 'World' } );
  Hello [% foo %]!
  END_TEMPLATE

=head1 DESCRIPTION

B<WARNING: THIS MODULE IS EXPERIMENTAL AND SUBJECT TO CHANGE WITHOUT NOTICE>

B<YOU HAVE BEEN WARNED!>

B<Template::Tiny> is a reimplementation of a partial subset of the
L<Template> Toolkit in as few lines of code as possible.

It is intended for use in light-usage, low-memory, or low-cpu templating
situations, where you may need to upgrade to the full feature set in the
future, or if you want the familiarity of TT-style templates.

It is intended to have fully-compatible template and stash usage,
with a limited by similar Perl API.

Unlike Template Toolkit, B<Template::Tiny> will process templates without a
compile phase (but despite this is still quicker, owing to heavy use of
the Perl regular expression engine.

=head2 SUPPORTED USAGE

Only the default C<[% %]> tag style is supported.

Both the [%+ +%] style explicit whitespace and the [%- -%] style explicit
chomp are support, although the [%+ +%] version is unneeded as Template::Tiny
does not support default-enabled PRE_CHOMP or POST_CHOMP.

Variable expressions in the form foo.bar.baz are supported.

Appropriate simple behaviours for ARRAY reference, HASH reference and objects
are supported, but not "VMethods" such as array lengths.

IF, ELSE and UNLESS conditions are supported, but only with simple foo.bar.baz
conditions.

Support for looping is available, in the most simple [% FOREACH item IN list %]
form.

All four IF/ELSE/UNLESS/FOREACH control structures are able to be nested to
arbitrary depth.

The treatment of C<_private> hash and method keys is compatible with Template
Toolkit, returning null or false rather than the actual content of the hash key
or method.

Anything beyond this is currently out of scope

=head1 METHODS

=head2 new

  my $template = Template::Tiny->new(
      TRIM => 1,
  );

The C<new> constructor is provided for compatibility with Template Toolkit.

The only parameter it currently supports is C<TRIM> (which removes leading
and trailing whitespace from processed templates).

Additional parameters can be provided without error, but will be ignored.

=head2 process

  $template->process( \$input, $vars );

The C<process> method is called to process a template. The firsts parameter
is a reference to a text string containing the template text. A reference
to a hash may be passed as the second parameter containing definitions of
template variables.

=head1 SUPPORT

Bugs should be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Template-Tiny>

For other issues, or commercial enhancement or support, contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 SEE ALSO

L<Config::Simple>

=head1 COPYRIGHT

Copyright 2009 - 2010 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
